subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/services/EnvironmentChangeTracker.kt

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configurationcache.services
+
+import org.gradle.configuration.internal.UserCodeApplicationContext
+import org.gradle.configurationcache.problems.PropertyTrace
+import org.gradle.configurationcache.problems.location
+import org.gradle.internal.service.scopes.Scopes
+import org.gradle.internal.service.scopes.ServiceScope
+import java.util.concurrent.ConcurrentHashMap
+
+
+/**
+ * The environment state that was mutated at the configuration phase and has to be restored before running a build from the cache.
+ */
+@ServiceScope(Scopes.BuildTree::class)
+class EnvironmentChangeTracker(private val userCodeApplicationContext: UserCodeApplicationContext) {
+    private
+    val mutatedSystemProperties = ConcurrentHashMap<Any, SystemPropertyChange>()
+
+    @Volatile
+    private
+    var systemPropertiesCleared = false
+
+    fun isSystemPropertyMutated(key: String): Boolean {
+        return systemPropertiesCleared || mutatedSystemProperties.containsKey(key)
+    }
+
+    fun loadFrom(storedState: CachedEnvironmentState) {
+        if (storedState.cleared) {
+            systemPropertiesCleared = true
```

Comment:
❓ is the internal state of the service actually used when loading from the cache?

Reply:
It isn't so far, I'm doing it just in case. Do you think state restore should be removed?

Reply:
Thanks, I've added that